Implemented b-state for unicodeobject

Eddy114514 opened the pull request:
Implemented b-state for unicodeobject by utilizing 4 bit of a int in unicodeobject->PyCompactUnicodeObject->PyASCIIObject->state

Implemented 2 macro for accessing and setting b-state

Eddy114514 committed:
Merge pull request #1 from softdevteam/migration

merge

Eddy114514 committed:
add bstate for string obj of str(unicode) and byte
for unicode hide bstate inside of unused bit of int
for byte create a subclass object pg_bytes with
bstate field

Eddy114514 committed:
just do unicode

nanjekyejoannah commented on Include/cpython/unicodeobject.h:
```diff
@@ -135,7 +135,8 @@ typedef struct {
         /* Padding to ensure that PyUnicode_DATA() is always aligned to
            4 bytes (see issue #19537 on m68k). */
-        unsigned int :25;
+        unsigned int bstate:4
```
We should have a different integer for Unicode, byte, undetermined, no?

Eddy114514 replied:
Thanks for pointing out. I think since I allocate 4 bits for this field, i can use different number to represent different state like 0->unsure, 1->byte, 2->Unicode. But I think your suggestion is better that it can be more clear and easier to access and modify.  Will fix!

Eddy114514 committed:
add bstate for bytesobject
add bstate for unicodeobject
add some useful macro
add logic and warning for unicodeobject.concat

ltratt commented on Objects/unicodeobject.c:
```diff
@@ -10681,42 +10682,84 @@ PyUnicode_Concat(PyObject *left, PyObject *right)
+            return NULL;
+        }
+    }
+    else{
```
Formatting has gone wonky here.

ltratt commented on Include/cpython/bytesobject.h:
```diff
@@ -14,6 +17,13 @@ typedef struct {
+#define PyBytes_GET_BSTATE(op)       (((PyBytesObject *)(op))->bstate)
+#define PyBytes_SET_BSTATE(op, val)  (((PyBytesObject *)(op))->bstate = (unsigned int)(val))
+
+#define PG_BSTATE_LOAD_BYTES(op_) \
```
I'm not really sure what the intention of this macro is, and the one call site doesn't immediately make it obvious to me. Can we provide a quick doc string?

ltratt commented on Include/cpython/unicodeobject.h:
```diff
@@ -135,10 +135,15 @@ typedef struct {
            4 bytes (see issue #19537 on m68k). */
-        unsigned int :25;
+        unsigned int bstate:4;
+        unsigned int :21;
```
This line has now been divorced from its doc string, so the "always aligned to 4 bytes" no longer makes sense.

That means that `bstate` needs lifting to a few lines earlier _and_ having a short doc string. In particular, it's not obvious to me why this needs 4 bits. Doesn't it only need 2?

ltratt commented on Include/cpython/unicodeobject.h:
```diff
@@ -160,6 +165,51 @@ typedef struct {
+unsure + byte = byte
+
+unicode + unsure = unicode
+unicode + byte = unicode
```
Right, so this is I think the big thing we need to consider (alongside `byte + unicode = byte`). In a sense, whenever this scenario happens it's "bad": the user is mixing things up in a way that won't work in Python 3.

We then have three choices: (1) we arbitrarily pick a winner (which is what the above logic does) (2) we revert back to "unsure" or (3) we have an additional "mixed" kind. I can see advantages to all three possibilities but it's hard to know which might be best. My probable criteria would be "which causes the fewest spurious warnings after the first warning?" I suspect we actually need to try this on some real code to know.

ltratt commented:
There's one major thing to consider above (https://github.com/softdevteam/pygrate3/pull/37#discussion_r2536876880) but in addition:

* I don't think we want to repeat the same macros (with minor variations) twice. Can we put these into one place?
* PRs like this need at least one test.

nanjekyejoannah commented:
Lets pause due to the `bstate` discussion.

ltratt commented:
FWIW I think this PR is heading in the right direction!

nanjekyejoannah commented:
Well we talked of not needing `bstate` in pygrate3 but yeah lets touch base for consensus in the meeting.

ltratt commented:
Closing as agreed on the call earlier.
